org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/marker/internal/ErrorMarkerLocation.java

```diff
@@ -0,0 +1,136 @@
+
+    private final IResource resource;
+    private final int lineNumber;
+
+    public ErrorMarkerLocation(IResource resource) {
+        this(resource, 0);
+    }
+
+    public ErrorMarkerLocation(IResource resource, int lineNumber) {
+        this.resource = resource;
+        this.lineNumber = lineNumber;
+    }
+
+    public IResource getResource() {
+        return this.resource;
+    }
+
+    public int getLineNumber() {
+        return this.lineNumber;
+    }
+
+    public static ErrorMarkerLocation findErrorLocation(GradleBuild gradleBuild, Throwable t) {
+        Matcher errorLocationInStackTrace = tryFindStackTraceErrorLocation(t);
+
+        if (errorLocationInStackTrace != null) {
+            IResource resource = tryFindWorkspaceFile(gradleBuild, new Path(errorLocationInStackTrace.group(1)));
+            if (resource != null) {
+                int lineNumber = Integer.parseInt(errorLocationInStackTrace.group(2));
+                return new ErrorMarkerLocation(resource, lineNumber);
+            }
+        }
+
+        return new ErrorMarkerLocation(calculateFallbackLocation(gradleBuild));
+    }
+
+    private static Matcher tryFindStackTraceErrorLocation(Throwable t) {
+        if (t == null || t.getMessage() == null) {
+            return null;
+        }
+
+        Matcher matcher = ERROR_LOCATION_PATTERN.matcher(t.getMessage());
+        if (!matcher.find()) {
+            return tryFindStackTraceErrorLocation(t.getCause());
+        }
+
+        return matcher;
+    }
+
+    private static IResource tryFindWorkspaceFile(GradleBuild gradleBuild, Path filePath) {
+        for (IProject project : getWorkspaceProjectsFor(gradleBuild)) {
+            IPath projectLocation = project.getLocation();
+            if (projectLocation.isPrefixOf(filePath)) {
+                IFile file = project.getFile(filePath.makeRelativeTo(projectLocation));
+                if (file.exists()) {
+                    return file;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private static IResource calculateFallbackLocation(GradleBuild gradleBuild) {
+        Optional<IProject> projectOrNull = CorePlugin.workspaceOperations().findProjectByLocation(gradleBuild.getBuildConfig().getRootProjectDirectory());
+        if (projectOrNull.isPresent()) {
+            IProject project = projectOrNull.get();
+            if (GradleProjectNature.isPresentOn(project)) {
+                PersistentModel persistentModel = CorePlugin.modelPersistence().loadModel(project);
+                if (persistentModel.isPresent()) {
+                    IFile buildScript = project.getFile(persistentModel.getbuildScriptPath());
+                    if (buildScript.exists()) {
+                        return buildScript;
+                    }
+                }
+            }
+            return project;
+        } else {
+            return ResourcesPlugin.getWorkspace().getRoot();
```

Comment:
⭕️ Do we ever clean these up?

Reply:
The cleanup starts at the workspace root: https://github.com/eclipse/buildship/pull/634/files#diff-1158f74e59739104ec0d6c6687e86ea2R38